File: 2022/11/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {
	// --- Constants ---
	private static final String INPUT_FILE = "../input/input.txt";
	private static final String PARSER_REGEX = "Monkey (\\d+):\\s+" +
		"Starting items: ([\\d, ]+)\\s+" +
		"Operation: new = old ([\\*\\+]) (\\d+|old)\\s+" +
		"Test: divisible by (\\d+)\\s+" +
		"If true: throw to monkey (\\d+)\\s+" +
		"If false: throw to monkey (\\d+)";

	public static void main(String[] args) throws IOException {
		List<Monkey> group1 = new ArrayList<>();
		List<Monkey> group2 = new ArrayList<>();

		// --- Read and parse the input file ---
		String text = Files.readString(Paths.get(INPUT_FILE));
		Pattern pattern = Pattern.compile(PARSER_REGEX);
		Matcher matcher = pattern.matcher(text);
		while (matcher.find()) {
			int id = Integer.parseInt(matcher.group(1));
			long[] items = Arrays.stream(matcher.group(2).split("\\s*,\\s*"))
				.mapToLong(Long::parseLong).toArray();
			char operator = matcher.group(3).charAt(0);
			String s = matcher.group(4);
			Integer operand = s.equals("old") ? null : Integer.valueOf(s);
			int divisible = Integer.parseInt(matcher.group(5));
			int throwTrue = Integer.parseInt(matcher.group(6));
			int throwFalse = Integer.parseInt(matcher.group(7));

			// --- Puzzle 1 ---
			Monkey monkey1 = new Monkey(id, items, operator, operand, divisible, throwTrue, throwFalse, 1, group1);
			group1.add(monkey1);

			// --- Puzzle 2 ---
			Monkey monkey2 = new Monkey(id, items, operator, operand, divisible, throwTrue, throwFalse, 2, group2);
			group2.add(monkey2);
		}

		// --- Puzzle 1 ---
		for (int i = 0; i < 20; i++) for (Monkey m: group1) m.play();
		long mb = group1.stream()
			.mapToLong(m -> m.getInspects())
			.boxed()
			.sorted(Comparator.reverseOrder())
			.limit(2)
			.reduce(1L, (a, b) -> a * b);
		System.out.format("1. Level of monkey business: %,d%n", mb);

		// --- Puzzle 2 ---
		for (int i = 0; i < 10_000; i++) for (Monkey m: group2) m.play();
		mb = group2.stream()
			.mapToLong(m -> m.getInspects())
			.boxed()
			.sorted(Comparator.reverseOrder())
			.limit(2)
			.reduce(1L, (a, b) -> a * b);
		System.out.format("2. Level of monkey business: %,d%n", mb);
	}
}

File: 2022/11/rust/src/main.rs
use regex::Regex;
use std::fmt::{ Display, Formatter };

// --- Constants --
const INPUT_FILE: &'static str = "../input/input.txt";
const PARSER_REGEX: &'static str = "Monkey (?:(\\d+)):\\s+\
        Starting items: (?:([\\d, ]+))\\s+\
        Operation: new = old (?:([\\+\\*])) (?:(\\d+|old))\\s+\
        Test: divisible by (?:(\\d+))\\s+\
        If true: throw to monkey (?:(\\d+))\\s+\
        If false: throw to monkey (?:(\\d+))";

// Represents a monkey in the game.
struct Monkey {
    id: usize,
    items: Vec<u64>,
    operator: char,
    operand: Option<u64>,
    divisible: u64,
    throw_true: usize,
    throw_false: usize,
    relief: usize,
    inspects: usize,
    gcd: u64
}

impl Monkey {
    fn new(id: usize, items: Vec<u64>, operator: char, operand: Option<u64>, divisible: u64, throw_true: usize, throw_false: usize, relief: usize) -> Self {
        Self { id, items, operator, operand, divisible, throw_true, throw_false, relief, inspects: 0, gcd: 0 }
    }

    fn play(&mut self) -> Vec<(usize, u64)> {
        let mut throws = Vec::new();
        while self.items.len() > 0 {
            self.inspects += 1;
            let mut item = self.items.remove(0);
            let value = if self.operand.is_none() { item } else { self.operand.unwrap() };
            match self.operator {
                '+' => item += value,
                '*' => item *= value,
                _ => panic!("Unsupported operator: {}", self.operator)
            }
            match self.relief {
                1 => item /= 3,
                2 => item %= self.gcd,
                _ => panic!("Unsupported relief: {}", self.relief)
            }
            let index = if item % self.divisible == 0 { self.throw_true } else { self.throw_false };
            throws.push((index, item));
        }
        throws
    }
}

impl Display for Monkey {
    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
        write!(f, "[Monkey] id: {},  items: {:?}, inspects: {}", self.id, self.items, self.inspects)
    }
}

fn parse_input_file(relief: usize) -> Result<Vec<Monkey>, Box<dyn std::error::Error>> {
    let mut group: Vec<Monkey> = Vec::new();
    let text = std::fs::read_to_string(INPUT_FILE)?;
    let regex = Regex::new(PARSER_REGEX)?;
    for cap in regex.captures_iter(&text) {
        let id: usize = cap[1].parse()?;
        let items: Vec<u64> = cap[2].split(",").map(|x| x.trim().parse::<u64>().unwrap()).collect();
        let operator: char = cap[3].chars().next().unwrap();
        let operand: Option<u64> = if &cap[4] == "old" { None } else { Some(cap[4].parse()?) };
        let divisible: u64 = cap[5].parse()?;
        let throw_true: usize = cap[6].parse()?;
        let throw_false: usize = cap[7].parse()?;
        let monkey = Monkey::new(id, items, operator, operand, divisible, throw_true, throw_false, relief);
        group.push(monkey);
    }
    Ok(group)
}

fn calculate_monkey_business(group: &Vec<Monkey>) -> u64 {
    let mut v = group.iter().map(|m| m.inspects).collect::<Vec<usize>>();
    v.sort();
    v.reverse();
    v.into_iter().take(2).fold(1, |acc, x| acc * x as u64)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // --- Puzzle 1 ---
    let mut group = parse_input_file(1)?;
    for _ in 0..20 {
        for i in 0..group.len() {
            let mut monkey = group.get_mut(i).unwrap();
            let throws = monkey.play();
            for (i, item) in throws.into_iter() {
                monkey = group.get_mut(i).unwrap();
                monkey.items.push(item);
            }
        }
    }
    let mb = calculate_monkey_business(&group);
    println!("1. Level of monkey business: {}", mb);

    // --- Puzzle 2 ---
    let mut group = parse_input_file(2)?;
    let gcd = group.iter().map(|m| m.divisible as u64).fold(1, |acc, x| acc * x);
    for monkey in group.iter_mut() {
        monkey.gcd = gcd;
    }

    for _ in 0..10_000 {
        for i in 0..group.len() {
            let mut monkey = group.get_mut(i).unwrap();
            let throws = monkey.play();
            for (i, item) in throws.into_iter() {
                monkey = group.get_mut(i).unwrap();
                monkey.items.push(item);
            }
        }
    }
    let mb = calculate_monkey_business(&group);
    println!("2. Level of monkey business: {}", mb);

    Ok(())
}
